src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedX509ExtensionProcessor.cs

```diff
@@ -128,90 +103,67 @@ public virtual void DecodeX509KeyUsageExtension(byte[] encoded, out X509KeyUsage
-            }
-            else
-            {
-                certificateAuthority = false;
-            }
-
-            if (constraints.HasData)
-            {
-                pathLengthConstraint = constraints.ReadInteger();
-                hasPathLengthConstraint = true;
-            }
-            else
-            {
-                pathLengthConstraint = 0;
-                hasPathLengthConstraint = false;
-            }
+            BasicConstraintsAsn constraints = AsnSerializer.Deserialize<BasicConstraintsAsn>(encoded, AsnEncodingRules.DER);
+            certificateAuthority = constraints.CA;
+            hasPathLengthConstraint = constraints.PathLengthConstraint.HasValue;
+            pathLengthConstraint = constraints.PathLengthConstraint.GetValueOrDefault();
         }
 
         public virtual byte[] EncodeX509EnhancedKeyUsageExtension(OidCollection usages)
         {
-            //extKeyUsage EXTENSION ::= {
-            //    SYNTAX SEQUENCE SIZE(1..MAX) OF KeyPurposeId
-            //    IDENTIFIED BY id-ce-extKeyUsage }
+            // https://tools.ietf.org/html/rfc5280#section-4.2.1.12
             //
-            //KeyPurposeId::= OBJECT IDENTIFIER
-
-            List<byte[][]> segments = new List<byte[][]>(usages.Count);
+            // extKeyUsage EXTENSION ::= {
+            //     SYNTAX SEQUENCE SIZE(1..MAX) OF KeyPurposeId
+            //     IDENTIFIED BY id-ce-extKeyUsage
+            // }
+            //
+            // KeyPurposeId ::= OBJECT IDENTIFIER
 
-            foreach (Oid usage in usages)
+            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
             {
-                segments.Add(DerEncoder.SegmentedEncodeOid(usage));
+                writer.PushSequence();
+                foreach (Oid usage in usages)
+                {
+                    writer.WriteObjectIdentifier(usage);
+                }
+                writer.PopSequence();
+                return writer.Encode();
             }
-
-            return DerEncoder.ConstructSequence(segments);
         }
 
         public virtual void DecodeX509EnhancedKeyUsageExtension(byte[] encoded, out OidCollection usages)
         {
+            // https://tools.ietf.org/html/rfc5924#section-4.1
+            //
             // ExtKeyUsageSyntax ::= SEQUENCE SIZE (1..MAX) OF KeyPurposeId
             //
-            // KeyPurposeId::= OBJECT IDENTIFIER
+            // KeyPurposeId ::= OBJECT IDENTIFIER
 
-            OidCollection oids = new OidCollection();
-            DerSequenceReader reader = new DerSequenceReader(encoded);
-
-            while (reader.HasData)
-            {
-                oids.Add(reader.ReadOid());
-            }
-
-            usages = oids;
+            Oid[] keyUsages = AsnSerializer.Deserialize<Oid[]>(encoded, AsnEncodingRules.DER);
+            usages = new OidCollection();
+            foreach (Oid KeyPurposeId in keyUsages)
+                usages.Add(KeyPurposeId);
```

Comment:
This is the second time I would really appreciate `OidCollection(ICollection<Oid>)` constructor. Oh well.